File: scripts/load_nflverse_data.py
#!/usr/bin/env python3
"""
NFLverse Data Loader (Python version)

Downloads player statistics from nflverse CSV files and loads into PostgreSQL.

Prerequisites:
    pip install pandas psycopg2-binary requests

Usage:
    python scripts/load_nflverse_data.py

Environment variables:
    DATABASE_URL - PostgreSQL connection string (Heroku format)
"""

import os
import sys
import pandas as pd
import psycopg2
from urllib.parse import urlparse

def parse_database_url(url):
    """Parse Heroku DATABASE_URL into connection parameters"""
    result = urlparse(url)
    return {
        'database': result.path[1:],
        'user': result.username,
        'password': result.password,
        'host': result.hostname,
        'port': result.port or 5432
    }

def get_db_connection():
    """Connect to PostgreSQL database"""
    database_url = os.getenv('DATABASE_URL')
    if not database_url:
        print("ERROR: DATABASE_URL environment variable not set")
        sys.exit(1)

    # Heroku uses postgres:// but psycopg2 expects postgresql://
    if database_url.startswith('postgres://'):
        database_url = database_url.replace('postgres://', 'postgresql://', 1)

    print("Connecting to database...")
    conn = psycopg2.connect(database_url, sslmode='require')
    print("Connected!\n")
    return conn

def fetch_nflverse_player_stats(season):
    """Download player stats CSV from nflverse GitHub"""
    url = f"https://github.com/nflverse/nflverse-data/releases/download/player_stats/player_stats_{season}.csv"
    print(f"Fetching player stats for {season} from nflverse...")

    try:
        df = pd.read_csv(url)
        print(f"  Downloaded {len(df)} records")
        return df
    except Exception as e:
        print(f"  ERROR: Failed to download: {e}")
        return None

def aggregate_season_stats(df):
    """Aggregate weekly stats into season totals per player"""
    print("Aggregating weekly stats into season totals...")

    # Group by player and aggregate stats
    agg_stats = df.groupby(['player_id', 'player_display_name', 'position', 'season', 'recent_team']).agg({
        'week': 'count',  # games played
        'passing_yards': 'sum',
        'passing_tds': 'sum',
        'interceptions': 'sum',
        'passing_attempts': 'sum',
        'completions': 'sum',
        'rushing_yards': 'sum',
        'rushing_tds': 'sum',
        'rushing_attempts': 'sum',
        'receiving_yards': 'sum',
        'receiving_tds': 'sum',
        'receptions': 'sum',
        'targets': 'sum'
    }).reset_index()

    # Rename columns
    agg_stats.rename(columns={
        'week': 'games',
        'player_display_name': 'player_name',
        'interceptions': 'passing_ints'
    }, inplace=True)

    print(f"  Aggregated to {len(agg_stats)} player-season records\n")
    return agg_stats

def get_team_mapping(conn):
    """Get mapping of team abbreviations to UUIDs"""
    cursor = conn.cursor()
    cursor.execute("SELECT id, abbreviation FROM teams")
    mapping = {row[1]: row[0] for row in cursor.fetchall()}
    cursor.close()
    return mapping

def get_player_mapping(conn):
    """Get mapping of player names to UUIDs"""
    cursor = conn.cursor()
    cursor.execute("SELECT id, LOWER(name) as name FROM players")
    mapping = {row[1]: row[0] for row in cursor.fetchall()}
    cursor.close()
    return mapping

def load_stats_to_database(conn, stats, team_mapping, player_mapping):
    """Load aggregated stats into player_career_stats table"""
    print("Loading stats into database...")

    cursor = conn.cursor()
    inserted = 0
    updated = 0
    failed = 0

    for _, row in stats.iterrows():
        # Find player by name (case-insensitive)
        player_name = row['player_name'].lower()
        player_id = player_mapping.get(player_name)

        if not player_id:
            # Try without suffixes like Jr., Sr., III
            simplified_name = player_name.split(',')[0].split(' jr')[0].split(' sr')[0].split(' iii')[0]
            player_id = player_mapping.get(simplified_name)

        if not player_id:
            print(f"  ⚠️  Player not found: {row['player_name']}")
            failed += 1
            continue

        # Get team ID
        team_abbr = row['recent_team']
        team_id = team_mapping.get(team_abbr)

        # Upsert career stats
        query = """
            INSERT INTO player_career_stats (
                player_id, season, team_id, games_played,
                passing_yards, passing_tds, passing_ints,
                passing_attempts, passing_completions,
                rushing_yards, rushing_tds, rushing_attempts,
                receiving_yards, receiving_tds, receptions, receiving_targets
            ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
            ON CONFLICT (player_id, season)
            DO UPDATE SET
                team_id = EXCLUDED.team_id,
                games_played = EXCLUDED.games_played,
                passing_yards = EXCLUDED.passing_yards,
                passing_tds = EXCLUDED.passing_tds,
                passing_ints = EXCLUDED.passing_ints,
                passing_attempts = EXCLUDED.passing_attempts,
                passing_completions = EXCLUDED.passing_completions,
                rushing_yards = EXCLUDED.rushing_yards,
                rushing_tds = EXCLUDED.rushing_tds,
                rushing_attempts = EXCLUDED.rushing_attempts,
                receiving_yards = EXCLUDED.receiving_yards,
                receiving_tds = EXCLUDED.receiving_tds,
                receptions = EXCLUDED.receptions,
                receiving_targets = EXCLUDED.receiving_targets
        """

        try:
            cursor.execute(query, (
                player_id, int(row['season']), team_id, int(row['games']),
                int(row['passing_yards']), int(row['passing_tds']), int(row['passing_ints']),
                int(row['passing_attempts']), int(row['completions']),
                int(row['rushing_yards']), int(row['rushing_tds']), int(row['rushing_attempts']),
                int(row['receiving_yards']), int(row['receiving_tds']),
                int(row['receptions']), int(row['targets'])
            ))

            if cursor.rowcount > 0:
                inserted += 1
            else:
                updated += 1

        except Exception as e:
            print(f"  ❌ Error for {row['player_name']}: {e}")
            failed += 1

    conn.commit()
    cursor.close()

    print(f"\n✅ Completed: {inserted} inserted, {updated} updated, {failed} failed\n")

def main():
    """Main execution"""
    print("NFLverse Data Loader (Python)")
    print("=" * 50)
    print()

    # Connect to database
    conn = get_db_connection()

    # Get mappings
    print("Loading team and player mappings...")
    team_mapping = get_team_mapping(conn)
    player_mapping = get_player_mapping(conn)
    print(f"  Found {len(team_mapping)} teams, {len(player_mapping)} players\n")

    # Process each season
    seasons = [2023, 2024]
    for season in seasons:
        print(f"\n{'='*50}")
        print(f"Processing {season} Season")
        print('='*50)

        # Fetch data
        df = fetch_nflverse_player_stats(season)
        if df is None:
            continue

        # Aggregate stats
        stats = aggregate_season_stats(df)

        # Load to database
        load_stats_to_database(conn, stats, team_mapping, player_mapping)

    # Close connection
    conn.close()
    print("\n✅ All done!\n")

if __name__ == '__main__':
    main()

File: scripts/load_nflverse_data.R
#!/usr/bin/env Rscript

# NFLverse Data Loader
# Fetches player statistics from nflverse and loads into PostgreSQL database
#
# Prerequisites:
#   install.packages(c("nflreadr", "DBI", "RPostgres", "dplyr", "tidyr"))
#
# Usage:
#   Rscript scripts/load_nflverse_data.R
#
# Environment variables:
#   DATABASE_URL - PostgreSQL connection string (Heroku format)

library(nflreadr)
library(DBI)
library(RPostgres)
library(dplyr)
library(tidyr)

# Parse Heroku DATABASE_URL
parse_database_url <- function(url) {
  # Format: postgres://user:password@host:port/database
  pattern <- "postgres://([^:]+):([^@]+)@([^:]+):([^/]+)/(.*)"
  matches <- regmatches(url, regexec(pattern, url))[[1]]

  if (length(matches) == 0) {
    stop("Invalid DATABASE_URL format")
  }

  list(
    user = matches[2],
    password = matches[3],
    host = matches[4],
    port = as.integer(matches[5]),
    dbname = matches[6]
  )
}

# Get database connection
get_db_connection <- function() {
  database_url <- Sys.getenv("DATABASE_URL")

  if (database_url == "") {
    stop("DATABASE_URL environment variable not set")
  }

  db_params <- parse_database_url(database_url)

  dbConnect(
    RPostgres::Postgres(),
    dbname = db_params$dbname,
    host = db_params$host,
    port = db_params$port,
    user = db_params$user,
    password = db_params$password,
    sslmode = "require"
  )
}

# Load player stats from nflverse
load_player_stats <- function(seasons = 2020:2024) {
  cat("Fetching player stats from nflverse...\n")

  # Load player stats - this uses nflverse's weekly stats
  stats <- load_player_stats(seasons) %>%
    group_by(player_id, player_name, position, season, recent_team) %>%
    summarise(
      games = n(),
      passing_yards = sum(passing_yards, na.rm = TRUE),
      passing_tds = sum(passing_tds, na.rm = TRUE),
      passing_ints = sum(interceptions, na.rm = TRUE),
      rushing_yards = sum(rushing_yards, na.rm = TRUE),
      rushing_tds = sum(rushing_tds, na.rm = TRUE),
      receiving_yards = sum(receiving_yards, na.rm = TRUE),
      receiving_tds = sum(receiving_tds, na.rm = TRUE),
      receptions = sum(receptions, na.rm = TRUE),
      targets = sum(targets, na.rm = TRUE),
      .groups = "drop"
    )

  cat(sprintf("Fetched %d player-season records\n", nrow(stats)))
  return(stats)
}

# Map nflverse team abbreviations to our database team IDs
get_team_mapping <- function(conn) {
  teams <- dbGetQuery(conn, "SELECT id, abbreviation FROM teams")
  setNames(teams$id, teams$abbreviation)
}

# Map nflverse player IDs to our database player UUIDs
get_player_mapping <- function(conn) {
  # Our database uses nfl_id which is the ESPN athlete ID
  # nflverse uses gsis_id as player_id
  # We need to match by name as a fallback
  players <- dbGetQuery(conn, "SELECT id, nfl_id, name FROM players")
  players
}

# Load career stats into database
load_career_stats_to_db <- function(conn, stats) {
  cat("Loading career stats into database...\n")

  team_mapping <- get_team_mapping(conn)
  players_db <- get_player_mapping(conn)

  inserted <- 0
  failed <- 0

  for (i in 1:nrow(stats)) {
    row <- stats[i, ]

    # Try to find player by name (fuzzy matching would be better)
    player_match <- players_db %>%
      filter(tolower(name) == tolower(row$player_name))

    if (nrow(player_match) == 0) {
      cat(sprintf("Player not found: %s\n", row$player_name))
      failed <- failed + 1
      next
    }

    player_id <- player_match$id[1]

    # Get team ID
    team_abbr <- row$recent_team
    team_id <- team_mapping[team_abbr]

    if (is.na(team_id) || is.null(team_id)) {
      team_id <- NULL
    }

    # Upsert career stats
    query <- "
      INSERT INTO player_career_stats (
        player_id, season, team_id, games_played,
        passing_yards, passing_tds, passing_ints,
        rushing_yards, rushing_tds,
        receiving_yards, receiving_tds, receptions, receiving_targets
      ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13)
      ON CONFLICT (player_id, season)
      DO UPDATE SET
        team_id = EXCLUDED.team_id,
        games_played = EXCLUDED.games_played,
        passing_yards = EXCLUDED.passing_yards,
        passing_tds = EXCLUDED.passing_tds,
        passing_ints = EXCLUDED.passing_ints,
        rushing_yards = EXCLUDED.rushing_yards,
        rushing_tds = EXCLUDED.rushing_tds,
        receiving_yards = EXCLUDED.receiving_yards,
        receiving_tds = EXCLUDED.receiving_tds,
        receptions = EXCLUDED.receptions,
        receiving_targets = EXCLUDED.receiving_targets
    "

    tryCatch({
      dbExecute(
        conn, query,
        params = list(
          player_id, row$season, team_id, row$games,
          row$passing_yards, row$passing_tds, row$passing_ints,
          row$rushing_yards, row$rushing_tds,
          row$receiving_yards, row$receiving_tds, row$receptions, row$targets
        )
      )
      inserted <- inserted + 1
    }, error = function(e) {
      cat(sprintf("Error inserting stats for %s: %s\n", row$player_name, e$message))
      failed <- failed + 1
    })
  }

  cat(sprintf("\nCompleted: %d inserted, %d failed\n", inserted, failed))
}

# Main execution
main <- function() {
  cat("NFLverse Data Loader\n")
  cat("===================\n\n")

  # Connect to database
  cat("Connecting to database...\n")
  conn <- get_db_connection()
  cat("Connected!\n\n")

  # Load stats from nflverse
  stats <- load_player_stats(seasons = 2023:2024)

  # Load into database
  load_career_stats_to_db(conn, stats)

  # Disconnect
  dbDisconnect(conn)
  cat("\nDone!\n")
}

# Run if called as script
if (!interactive()) {
  main()
}
